### PR TITLE
yes

### PR DESCRIPTION
**Pull request by dannykha**
thumbs up

### CODE REVIEW

File: homework2.iml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="false" />
+      <sourceFolder url="file://$MODULE_DIR$/src/test/java" isTestSource="true" />
+      <excludeFolder url="file://$MODULE_DIR$/target" />
+    </content>
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+    <orderEntry type="library" name="Maven: junit:junit:4.13" level="project" />
+    <orderEntry type="library" name="Maven: org.hamcrest:hamcrest-core:1.3" level="project" />
+  </component>
+</module>
```
Comment (pdgetrf) [inline homework2.iml:16]: do not check in iml file

File: src/main/java/Problem1.java [hunk 2]
```diff
@@ -2,9 +2,42 @@
...
     public static void colorSort(int[] nums) {
+        int low = 0;
+        int mid = 0;
+        int temp = 0;
+        int high = nums.length - 1;
+        while (mid <= high) {
+            if (nums[mid] == 0) {
+                temp = nums[low];
+                nums[low] = nums[mid];
+                nums[mid] = temp;
```
Comment (pdgetrf) [inline src/main/java/Problem1.java:30]: I see repeated logic here. refactor the swap into a private function. 

we prefer single source of truth.

File: src/main/java/Problem3/Book.java [hunk 3]
```diff
@@ -16,10 +16,19 @@ public Book(String title, String author) {
...
+        this.author = anotherBook.author;
+        this.id = anotherBook.id;
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        Book book = (Book)obj;
+        if (book.title == this.title  && book.author == this.author
+                && book.id == this.id) {
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:29]: wrong logic 

```
A movie equals another movie only when their id are the same. Same rule for books being equal.
```

File: src/main/java/Problem3/BookFiction.java [hunk 4]
```diff
@@ -6,14 +6,16 @@
 
     private String genres;
 
+
     public BookFiction(String title, String author, String genres) {
-        // homework
+        super(title, author);
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookFiction.java:11]: what about genres?

File: src/main/java/Problem3/Movie.java [hunk 5]
```diff
@@ -9,15 +9,27 @@
...
+        rating = anotherMovie.rating;
+        title = anotherMovie.title;
         // homework
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        Movie movie = (Movie)obj;
+        if (movie.title == this.title && movie.rating == this.rating
```
Comment (pdgetrf) [inline src/main/java/Problem3/Movie.java:28]: same here. wrong logic.

File: src/main/java/Problem3/MovieAction.java [hunk 6]
```diff
@@ -21,6 +23,12 @@ public int getLateFeeInDollar() {
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+        if (numOfDaysPastDue < 5 && numOfDaysPastDue > 0) {
+            return (numOfDaysPastDue * lateFeePerDayInDollar);
+        }
+        else if (numOfDaysPastDue >= 5) {
+            return (2 * numOfDaysPastDue * lateFeePerDayInDollar);
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:30]: also duplicated logic here. 

suggesting:
```java
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

and here number 5 only appears once too

File: src/test/java/Problem2Test.java [hunk 7]
```diff
@@ -7,13 +7,23 @@
...
         int[][] expects = {
-                {0, 1, 2, 4, 6}
-                // add more test cases
+                {0, 1, 2, 4, 6},
+                {0},
+                {101, 102, 103, 104, 105},
+                {56, 109, 2039, 5845},
+                {1, 12, 123, 1234, 12345},
+                {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+                {0, 1, 2, 3}
```
Comment (pdgetrf) [inline src/test/java/Problem2Test.java:26]: great test. let's add more cases for the edge cases, such as null array and empty array.

**Review by pdgetrf — APPROVED** [review]
LGTM
